### ISSUE TITLE
404 "The requested URL /firefly-iii/public/login was not found on this server."

### ISSUE THREAD

**Issue by zimmertr** [JC5/firefly-iii#317]
I followed your guide for Ubuntu 16.04 and I'm unable to get it to work. Additionally, in apache2.conf, I had to modify the "allow for all" directive to "allow from all" otherwise the apache service would fail to start.

When I try to navigate to 

localhost/firefly-iii/public 

it redirects to 

localhost/firefly-iii/public/login

which is a 404 page saying that "The requested URL /firefly-iii/public/login was not found on this server."

Have I done something wrong?

**Comment by zimmertr** [comment]
I also had to manually log into mysql and create a database. Unsure if it's intended to be automatically created or not.

**Comment by applelam** [comment]
It seems that you didn't follow the "Extra steps for Apache" on the installation guide. 
> I also had to manually log into mysql and create a database. Unsure if it's intended to be automatically created or not.
Yes, I believe it is intended for you to create a database manually, for security concern.

**Comment by zimmertr** [comment]
Thank you for responding.

Sure, here is my entire config file: http://pastebin.com/3vgPGLVZ

It appears that all 'AllowOverride' directives have been set to All and the the Directory section has been added. In it you'll see that I changed `allow for all` to `allow from all` because if I don't Apache refuses to start: http://pastebin.com/yDrkB046

**Comment by applelam** [comment]
Hmm.. the apache config seems okay to me. 
Did you enabled the rewrite module on apache? Enable it by: 

> # a2enmod rewrite

**Comment by zimmertr** [comment]
That wasn't enabled so that's probably a contributing factor. But I'm still experiencing the same issue. Here is an imgur album demonstrating the problem. In it you can see that I attempted to connect via http and https, and you can see the apache server running in another to show that the server works. 

Do I have to enable any other modules, maybe?

**Comment by applelam** [comment]
Hmm, you might need to install and enable a few PHP modules for firefly-iii. Here's a list of them: https://jc5.github.io/firefly-iii/system-requirements/ .

The installation guide includes commands to help you install all the PHP modules required. It said it is not necessary for Ubuntu 16 but I am not sure about it.

> sudo apt-get install -y php7.0 php7.0-cli php7.0-common php7.0-fpm php7.0-mysql php7.0-curl php7.0-gd php7.0-imap php7.0-intl php7.0-json php7.0-mcrypt php7.0-readline php7.0-tidy php7.0-zip php7.0-bcmath php7.0-xml php7.0-mbstring php7.0-sqlite php7.0-bz2

If you found the problem still persist after install those modules, please take a look at the log files. You may find firefly-iii log files in ```firefly-iii/storage/logs``` and apache error log: ```/var/log/apache2/error.log```

**Comment by JC5** [comment]
Is the directory /storage/ writable for your web server?

**Comment by zimmertr** [comment]
Some of those packages were missing but after installing them I'm still getting a 500 server error on the page. I didn't have time to parse through the log after work today but here it is on pastebin: http://pastebin.com/sjVxsMBp

And yes the directory is writable.

**Comment by JC5** [comment]
The top line seems to explain it.

``Access denied for user 'homestead'@'localhost' (using password: YES)``

Make sure that your database exists and that you configure Firefly (in the ``.env``) file to use the correct credentials.

**Comment by zimmertr** [comment]
This looks correct to me. The password I use to log into the database in the command where I connect to mysql as root is the one that used to be where "redacted" is in the output of the first command. 

```
root@thebe:/var/www/firefly-iii# cat .env | grep DB
DB_CONNECTION=mysql
DB_HOST=localhost
DB_DATABASE=homestead
DB_USERNAME=root
DB_PASSWORD=redacted

root@thebe:/var/www/firefly-iii# mysql -u root -p
Enter password:
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 34
Server version: 5.7.15-0ubuntu0.16.04.1 (Ubuntu)

Copyright (c) 2000, 2016, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> show databases;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| homestead          |
| mysql              |
| performance_schema |
| sys                |
+--------------------+
5 rows in set (0.00 sec)
```

Does something look amiss to you?

**Comment by JC5** [comment]
If the log stays empty, the map may not be writeable by your web server.

Try changing the DB_HOST setting to 127.0.0.1

Otherwise, I'm as lost as you are.

**Comment by zimmertr** [comment]
Unfortunately the file is writable and is still blank: 

```
root@thebe:/var/www/firefly-iii/storage/logs# cat firefly-iii-2016-09-18.log

root@thebe:/var/www/firefly-iii/storage/logs# ls -lah
total 16K
drwxrwxr-x  2 www-data www-data 4.0K Sep 18 14:24 .
drwxrwxrwx 10 www-data www-data 4.0K Sep 20 14:32 ..
-rw-rw-rw-  1 www-data www-data    1 Sep 20 10:17 firefly-iii-2016-09-18.log
-rw-rw-r--  1 www-data www-data   14 Sep 18 14:21 .gitignore
```

Also, updating the .env for using 127.0.0.1 as the host did not work either. Still receiving a 500 page. :(

**Comment by JC5** [comment]
It still gives the "no mysql access" error? What do the Apache logs show you? There should be an entry for each 500 hit.

**Comment by zimmertr** [comment]
Access.log is filled with this line:

`127.0.0.1 - - [21/Sep/2016:09:22:38 -0400] "GET /server-status?auto HTTP/1.1" 200 960 "-" "Python-urllib/2.7"`

Error.log is has these two lines:

```
[Wed Sep 21 06:25:03.409425 2016] [mpm_prefork:notice] [pid 1338] AH00163: Apache/2.4.18 (Ubuntu) configured -- resuming normal operations
[Wed Sep 21 06:25:03.409483 2016] [core:notice] [pid 1338] AH00094: Command line: '/usr/sbin/apache2'
```

other_vhosts_access.log is empty.

**Comment by applelam** [comment]
Hmm. I believe you have modified ```/var/www/firefly-iii/.env``` with your database credentials but firefly's config file actually lives in ```/var/www/html/firefly-iii/.env```.

This explain why your ```.env``` file have ```DB_USERNAME=root``` but the log shows ```Access denied for user 'homestead'@'localhost' (using password: YES)```, where username is ```homestead``` instead of ```root```.

The apache2 log reveals the actual location of Firefly III.
```
PHP Fatal error:  Uncaught UnexpectedValueException: The stream or file "/var/www/html/firefly-iii/storage/logs/firefly-iii-2016-09-20.log" could not be opened
```

**Comment by JC5** [comment]
Good call.

**Comment by zimmertr** [comment]
I just ran a vimdiff on /var/www/firefly-iii/.env and /var/www/html/firefly-iii/.env and the only difference was for DB_HOST. The former was set to localhost and the later was set to 127.0.0.1.

That being said, I changed the ownership of /var/www/html/firefly/ to www-data like the other directory and it started working. Thanks for all your help guys! I've never had such awesome support on a Github issue before.
